Chapter user results when watching a group

GabrielVidal1 opened the pull request:
Fixes #216
Add the content of the Chapter View table in the Participation tab when looking at a chapter and watching a group at the same time.

smadbe reviewed (changes requested):
The layout when loading is not clean. missing space under the tabs, line on the half of page, ... 
![Screenshot 2020-11-25 at 18 21 41](https://user-images.githubusercontent.com/1053150/100261520-241f0f00-2f4b-11eb-8e78-87e4175b049f.png)

The table is also not really as expected:
- no/wrong font on leftmost column
- white cell not as expected 
- many lines while there should not be that many
![Screenshot 2020-11-25 at 18 24 10](https://user-images.githubusercontent.com/1053150/100261806-97288580-2f4b-11eb-8f43-57236b5c8ade.png)
![Screenshot 2020-11-25 at 18 24 22](https://user-images.githubusercontent.com/1053150/100261810-9859b280-2f4b-11eb-93cd-3b0c0e1358e3.png)

smadbe commented:
> Fixes #224
I think it is #216 instead of #224

GabrielVidal1 commented:
I fixed the css.
When the user name is too long, it will be clipped so I added a tooltip to show the full user name (`firstname lastname (login)`) when the cursor is above the name for at least 1 second.

smadbe commented:
> I fixed the css.
Nice. Could you make the the vertical line lighter as in the design as well?

> When the user name is too long, it will be clipped so I added a tooltip to show the full user name (`firstname lastname (login)`) when the cursor is above the name for at least 1 second.
👍   It would probably useful to do the same with the column header.

smadbe commented on src/app/shared/http-services/get-group-user-descendants.service.ts:
```diff
@@ -0,0 +1,66 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class GetGroupUserDescendantsService {
```
Maybe you should group all the descendant services in a single file. No need to have one service per file.

GabrielVidal1 replied:
On this branch I only have the `GetGroupUserDescendantsService` (the `GetGroupTeamDescendantsService` service was merged on master after I created this branch) so I will do this after the reviews, when I will rebase this branch on master.

smadbe commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -0,0 +1,81 @@
+import { GetItemChildrenService, ItemChild } from '../../../http-services/get-item-children.service';
+import { ItemData } from '../../../services/item-datasource.service';
+
+interface User {
```
I think it would be better to make `UserInfos` from UserPipe public (that I would call "FormattableUser") and use it here instead of redefining the same.

GabrielVidal1 replied:
fixed

smadbe commented on src/app/shared/http-services/get-group-users-progress.service.ts:
```diff
@@ -0,0 +1,56 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class GetGroupUsersProgressService {
```
Maybe you should group all the progress services in a single file.

GabrielVidal1 replied:
for now I just have one but I'll do this with the others.

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html:
```diff
@@ -0,0 +1,44 @@
+  >
+    <ng-template pTemplate="header" let-columns>
+      <tr style="border: none;">
+        <th style="min-width: 160px;transform: none;width: 159px;"></th>
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html:
```diff
@@ -0,0 +1,44 @@
+      let-rowIndex="rowIndex"
+    >
+      <tr>
+        <td class="users" pTooltip="{{ data.users[rowIndex] | user }}" tooltipPosition="left" [showDelay]="1000">
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.scss:
```diff
@@ -0,0 +1,73 @@
+
+  font-family: 'titillium', sans-serif;
+
+  tr > th {
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html:
```diff
@@ -0,0 +1,44 @@
+        </td>
+        <td *ngFor="let col of columns; let index = index">
+          <alg-user-progress
+            [userProgress]="data.data[rowIndex][index]"
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -0,0 +1,81 @@
+    private getGroupUsersProgressService: GetGroupUsersProgressService,
+  ) { }
+
+  getData(itemId: string, groupId: string, attemptId: string): Observable<Data> {
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/user-progress/user-progress.component.html:
```diff
@@ -0,0 +1,16 @@
+<div
```
fixed

GabrielVidal1 commented on src/app/modules/shared-components/components/score-ring/score-ring.component.ts:
```diff
@@ -103,6 +104,8 @@ export class ScoreRingComponent implements OnInit, OnChanges {
       this.icon = undefined;
     }
 
+    if (this.textColor) this.textFill = this.textColor as ScoreRingColor;
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  ngOnChanges(_changes: SimpleChanges): void {
+    if (!this.itemData || !this.itemData.currentResult) return;
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -0,0 +1,81 @@
+
+    merge(
+      of(fetchingState()),
+      this.getData(this.itemData.item.id, this.group.id, this.itemData.currentResult.attemptId).pipe(map(readyState))
```
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -0,0 +1,81 @@
+})
+export class GroupSituationChapterViewComponent implements OnChanges {
+
+  @Input() group!: Group;
```
fixed

GabrielVidal1 commented on src/app/modules/item/pages/item-progress/item-progress.component.ts:
```diff
@@ -10,4 +12,7 @@ export class ItemProgressComponent {
 
   @Input() itemData?: ItemData;
 
+  session$ = this.sessionService.session$.pipe(delay(0));
```
I thought it would be also used for an `item-log-view` component when watching a group.
![image](https://user-images.githubusercontent.com/31007874/100633087-e98fea80-332d-11eb-8ff9-eac968a8d5cd.png)

smadbe replied:
Sessions (info about the current info and managed group) are something every component can access, so I don't see on what basis we would pass them by arg here and not at other places... That is not bad, but looking for consistency, I don't see reason to do that in this case.

GabrielVidal1 replied:
fixed

GabrielVidal1 commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -0,0 +1,81 @@
+      },
+      _err => {
+        this.state = 'error';
+      }
```
I kept the current structure.

smadbe commented on src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:
```diff
@@ -33,19 +33,20 @@ export class GroupSituationChapterViewComponent implements OnChanges {
     data: [],
   }
 
+  private dataFecthing = new Subject<{ groupId: string, itemId: string, attemptId: string }>();
```
typo on `dataFecthing`

GabrielVidal1 replied:
fixed

smadbe reviewed (changes requested):
👍 Big change required! ;-)
'+ the comment about passing session
